### PR TITLE
test: Add Atlas region override and OUT_OF_CAPACITY fallback to acceptance test helpers

### PR DESCRIPTION
**Pull request by oarbusi**
## Description

Part of CLOUDP-401312 (Acceptance Tests Smarter Region Selection).

Makes the Atlas regions used by acceptance test helpers configurable and adds automatic fallback on `OUT_OF_CAPACITY` errors, to reduce nightly suite failures in cloud-dev caused by region capacity constraints.

**Changes:**
- New `MONGODB_ATLAS_TEST_REGIONS` env var (comma-separated Atlas regions). First entry is the default region, the rest are fallback candidates. Defaults to `US_WEST_2,US_EAST_1,US_EAST_2` when unset.
- `acc.DefaultRegions()` / `acc.DefaultRegion()` helpers in `internal/testutil/acc/region.go`.
- Default-region injection points now use the helper: `ReplicationSpecRequest.AddDefaults()`, shared execution cluster `clusterReq()`, `ConfigBasicDedicated` and `ConfigDedicatedNVMeBackupEnabled`.
- `createCluster()` (shared execution cluster, SDK-created) retries across candidate regions when creation fails with `OUT_OF_CAPACITY` or provisioning ends in a terminal unexpected state, deleting the broken cluster before the next attempt (a failed delete fails the test). Other errors still fail immediately.
- Unit tests for the fallback loop in `internal/testutil/acc/atlas_test.go` (black box, via an `export_test.go` shim).
- `acceptance-tests-runner.yml` exposes the repo variable `MONGODB_ATLAS_TEST_REGIONS` to all test jobs, so future capacity issues can be mitigated by editing a GitHub variable instead of merging code.

**Notes:**
- Test-only change: no provider/resource code touched.
- Fallback only covers SDK-created clusters (shared execution cluster). Clusters created via Terraform configs cannot retry at runtime (terraform-plugin-testing has no hook for it) and still rely on the region override.
- The Atlas capacity admin endpoints (`/api/private/nds/cloudProviders/{provider}/denyListEntries`) require `GLOBAL_ADMIN_READ_ONLY`, so they cannot be queried with the current cloud-dev CI key; denylist-aware selection is a possible follow-up if that role is granted.

## How to test

1. Dispatch `acceptance-tests.yml` with `ref: CLOUDP-401312`, `test_group: search_index` (uses the shared execution cluster): green run proves no regression in the default path. ✅ [Success](https://github.com/mongodb/terraform-provider-mongodbatlas/actions/runs/32230394906)
2. Set repo variable `MONGODB_ATLAS_TEST_REGIONS` to a single valid region (e.g. `EU_WEST_1`) and re-dispatch: proves CI plumbing. ✅ [Success](https://github.com/mongodb/terraform-provider-mongodbatlas/actions/runs/32235508406/job/96014494434), [Success](https://github.com/mongodb/terraform-provider-mongodbatlas/actions/runs/32251715081)
3. For an end-to-end fallback check, once OUT_OF_CAPACITY has been reached, trigger more test: job logs should show `OUT_OF_CAPACITY, trying next region` and tests should pass.

Locally: `go test ./internal/testutil/acc/` passes with the env var unset and set; `golangci-lint` and `go fix -diff` are clean.

### CODE REVIEW

File: internal/testutil/acc/atlas.go [hunk 1]
```diff
@@ -41,16 +42,36 @@ func deleteProject(id string) error {
...
-	_, err = stateConf.WaitForStateContext(tb.Context())
-	require.NoError(tb, err, "Cluster creation failed: %s, err: %s", name, err)
-
+	regions := DefaultRegions()
+	for i, region := range regions {
+		lastRegion := i == len(regions)-1
+		req := clusterReq(name, projectID, region, backupEnabled, pitEnabled)
+		_, _, err := ConnV2().ClustersAPI.CreateCluster(tb.Context(), projectID, &req).Execute()
+		if isOutOfCapacityError(err) && !lastRegion {
+			tb.Logf("Cluster creation in %s failed with OUT_OF_CAPACITY, trying next region: %s, err: %s", region, name, err)
```
[the remark this answers is not in the document]
Comment (EspenAlbert) [reply internal/testutil/acc/atlas.go]: 🤖 confirmed at head `cde5b6756`: both `tb.Logf` calls at `internal/testutil/acc/atlas.go:51` and `:58` pass `name` in the `trying next region: %s` slot, so the cluster name is printed where the next candidate region belongs. Easiest fix is to log the next region, e.g. `regions[i+1]`, so a capacity fallback shows which region the next attempt uses

File: internal/testutil/acc/atlas.go [hunk 2]
```diff
@@ -41,16 +42,36 @@ func deleteProject(id string) error {
...
+			continue
+		}
+		require.NoError(tb, err, "Cluster creation failed: %s, err: %s", name, err)
+		stateConf := cluster.CreateStateChangeConfig(tb.Context(), ConnV2(), projectID, name, 1*time.Hour)
+		_, err = stateConf.WaitForStateContext(tb.Context())
```
[the remark this answers is not in the document]
Comment (EspenAlbert) [reply internal/testutil/acc/atlas.go]: 🤖 confirmed, and the mechanism is a bit broader than described: `resourceRefreshFunc` (`internal/service/cluster/new_atlas.go:80-103`) returns only `cluster.GetStateName()` and discards the cluster `error` detail, so a terminal `FAILED` state surfaces from `WaitForStateContext` as `unexpected state 'FAILED', wanted target 'IDLE'` (terraform-plugin-sdk v2.40.1 `helper/retry/error.go:45`) with no `OUT_OF_CAPACITY` text. `isOutOfCapacityError` on that error is false, so the provisioning-time `continue` at `atlas.go:62` is unreachable unless the `GetCluster` call itself errors with `OUT_OF_CAPACITY`. Suggested fix: on a non-last-region wait failure, call `GetCluster` and check the cluster's `error` field for the capacity detail instead of relying on the retry error string

File: internal/testutil/acc/atlas.go [hunk 3]
```diff
@@ -41,16 +42,36 @@ func deleteProject(id string) error {
...
+		if isOutOfCapacityError(err) && !lastRegion {
+			tb.Logf("Cluster creation in %s failed with OUT_OF_CAPACITY while provisioning, trying next region: %s, err: %s", region, name, err)
+			if err := deleteCluster(projectID, name); err != nil {
+				tb.Logf("Failed to delete cluster %s after OUT_OF_CAPACITY: %s", name, err)
+			}
+			continue
```
[the remark this answers is not in the document]
Comment (EspenAlbert) [reply internal/testutil/acc/atlas.go:99]: 🤖 confirmed: `atlas.go:59-61` swallows the `deleteCluster` error with `tb.Logf` and continues, and `deleteCluster` itself (`atlas.go:77-88`) waits up to 1h for `DELETED`, so a delete wait timeout hits the same path. The next `CreateCluster` reuses `name` (`atlas.go:49`), Atlas returns a name conflict that is not `OUT_OF_CAPACITY`, and `require.NoError` at `atlas.go:54` fails the test, so the remaining candidate regions are never tried. Two options: (1) make a failed delete fatal and surface both the provisioning and delete errors, or (2) use a fresh cluster name per attempt and return the name that succeeded

File: internal/testutil/acc/atlas.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (EspenAlbert) [reply internal/testutil/acc/atlas.go]: fixed at `ff63a850d`: both log calls now read `trying next region, err: %s` formatted with `(region, err)` (`atlas.go:53` and `:60`), so no placeholder prints the cluster name in the region slot

File: internal/testutil/acc/atlas.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (EspenAlbert) [reply internal/testutil/acc/atlas.go]: fixed at `ff63a850d`: `isFailedProvisioning` (`atlas.go:75-84`) retries on any `*retry.UnexpectedStateError` in addition to `OUT_OF_CAPACITY`, and the SDK returns that type unwrapped (`helper/retry/state.go:175`), so a terminal `FAILED` state now reaches the delete-and-continue fallback. This also retries on non-capacity terminal states, which the doc comment presents as a deliberate trade for test reliability; left one follow-up on unit coverage for the new predicate in a new thread

File: internal/testutil/acc/atlas.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (EspenAlbert) [reply internal/testutil/acc/atlas.go]: fixed at `ff63a850d`: the cleanup delete is now fatal via `require.NoError(tb, deleteCluster(projectID, name), "Cluster deletion failed: %s", name)` (`atlas.go:61`), so the loop never recreates with a name still owned by a failed or deleting cluster, matching option (1)

File: internal/testutil/acc/atlas.go [hunk 4]
```diff
@@ -41,16 +44,49 @@ func deleteProject(id string) error {
...
+	}
+	require.Fail(tb, "Cluster creation failed in all candidate regions", "name: %s, regions: %s", name, strings.Join(regions, ","))
 	return name
 }
 
+// isFailedProvisioning reports whether waiting for cluster creation failed in a way that
+// justifies trying another region: an explicit OUT_OF_CAPACITY error, or a terminal
+// unexpected cluster state reported by the polling endpoint (whose error would not
+// contain OUT_OF_CAPACITY text).
+func isFailedProvisioning(err error) bool {
```
Comment (EspenAlbert) [inline internal/testutil/acc/atlas.go]: `isFailedProvisioning` is the decision point the previous revision got wrong (the old predicate missed terminal `FAILED` states), and it lands without unit coverage. Add a table test covering nil, a plain error, an `OUT_OF_CAPACITY` error, a `*retry.UnexpectedStateError`, and a `*retry.TimeoutError` (plus a wrapped variant) so the retry semantics stay locked in; the `ErrIfSampleDatasetLoadFailedForTest` pattern (`advanced_cluster.go:198`) works for reaching the unexported helper from `acc_test`

File: internal/testutil/acc/atlas.go [hunk 5]
```diff
@@ -41,16 +44,49 @@ func deleteProject(id string) error {
...
 	tb.Helper()
-	req := clusterReq(name, projectID, backupEnabled, pitEnabled)
-	_, _, err := ConnV2().ClustersAPI.CreateCluster(tb.Context(), projectID, &req).Execute()
-	require.NoError(tb, err, "Cluster creation failed: %s, err: %s", name, err)
-	stateConf := cluster.CreateStateChangeConfig(tb.Context(), ConnV2(), projectID, name, 1*time.Hour)
-	_, err = stateConf.WaitForStateContext(tb.Context())
-	require.NoError(tb, err, "Cluster creation failed: %s, err: %s", name, err)
-
+	regions := DefaultRegions()
+	for i, region := range regions {
```
Comment (lantoli) [inline internal/testutil/acc/atlas.go:79]: this loop is the core behavior of the change, but the added tests only exercise region parsing,  consider adding more tests, e.g. deterministic coverage for create-time fallback, provisioning-time cleanup, unrelated errors, delete failures, ...

File: internal/testutil/acc/atlas.go [hunk 6]
```diff
@@ -41,16 +44,49 @@ func deleteProject(id string) error {
...
+	require.Fail(tb, "Cluster creation failed in all candidate regions", "name: %s, regions: %s", name, strings.Join(regions, ","))
 	return name
 }
 
+// isFailedProvisioning reports whether waiting for cluster creation failed in a way that
+// justifies trying another region: an explicit OUT_OF_CAPACITY error, or a terminal
+// unexpected cluster state reported by the polling endpoint (whose error would not
+// contain OUT_OF_CAPACITY text).
+func isFailedProvisioning(err error) bool {
+	if err == nil {
```
Comment (lantoli) [inline internal/testutil/acc/atlas.go]: "err == nil" is a redundant check, already covered in isOutOfCapacityError

File: internal/testutil/acc/atlas.go (hunk 5 above)
Comment (oarbusi) [reply internal/testutil/acc/atlas.go:79]: Added: the loop is extracted into `createClusterWithRegionFallback` with injected create/wait/delete funcs, and `atlas_test.go` now has table-driven coverage for create-time fallback, provisioning-time cleanup, unexpected states, unrelated errors, and delete failures. Kept black box per testing-best-practices via an `export_test.go` shim.

File: internal/testutil/acc/atlas.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (oarbusi) [reply internal/testutil/acc/atlas.go:99]: Fixed with option (1): a failed delete is now fatal and surfaces both the delete and provisioning errors, so the next region attempt cannot hit a same-name conflict. Covered by a unit test.

File: internal/testutil/acc/atlas.go (hunk 6 above)
Comment (oarbusi) [reply internal/testutil/acc/atlas.go]: Removed.

File: internal/testutil/acc/atlas.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (oarbusi) [reply internal/testutil/acc/atlas.go]: Fixed: both fallback log messages now print the actual next region via `regions[i+1]` instead of the cluster name.

File: internal/testutil/acc/atlas.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (oarbusi) [reply internal/testutil/acc/atlas.go]: Fixed: the provisioning wait now also falls back on `retry.UnexpectedStateError`, so a terminal cluster state surfaces no OUT_OF_CAPACITY text but still triggers the region fallback. Covered by a unit test.

File: internal/testutil/acc/atlas.go (hunk 5 above)
Comment (EspenAlbert) [reply internal/testutil/acc/atlas.go]: verified at `726393250`: the extracted `createClusterWithRegionFallback` (`atlas.go:76-108`) with injected create/wait/delete funcs is covered end to end in `atlas_test.go`: create-time fallback for both `OUT_OF_CAPACITY` shapes, provisioning-time cleanup with delete counting, unexpected-state fallback, unrelated create/wait errors failing fast, delete failure failing fast, and the all-regions-exhausted combined error; `go vet ./internal/testutil/acc/` and `go test ./internal/testutil/acc/` pass. The `export_test.go` shim also matches the existing `internal/config/export_test.go` convention from #4597, so no concern there

File: internal/testutil/acc/atlas_test.go [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+				assert.Contains(t, err.Error(), tc.wantErr)
+			}
+			assert.Equal(t, tc.wantCreated, tracker.createdRegions)
+			assert.Equal(t, tc.wantDeletes, tracker.deletes)
+		})
+	}
+}
+
+func TestIsFailedProvisioning(t *testing.T) {
+	assert.False(t, acc.IsFailedProvisioning(nil))
```
Comment (EspenAlbert) [inline internal/testutil/acc/atlas_test.go]: nit: `TestIsFailedProvisioning` covers nil, plain, wrapped, `OUT_OF_CAPACITY` (string and API error code), and `*retry.UnexpectedStateError`, but not `*retry.TimeoutError`, the other error type `WaitForStateContext` returns. Add `assert.False(t, acc.IsFailedProvisioning(&retry.TimeoutError{}))` so a poll timeout is pinned as fail-fast rather than region fallback

File: internal/testutil/acc/region.go [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+		}
+	}
+	if len(result) == 0 {
+		return slices.Clone(defaultRegions)
+	}
+	return result
+}
+
+// DefaultRegion returns the Atlas region used by default in acceptance test helpers.
+func DefaultRegion() string {
```
Comment (EspenAlbert) [inline internal/testutil/acc/region.go:39]: Nice. Maybe we should try to use this in more places?
Even using DefaultRegion2 or DefaultRegion3 in places where multiple regions are necessary instead of hard-coding

**Review by EspenAlbert — APPROVED** [review]
Nice. Curious if you tried to read capacity from the API so we could pick based on that?

**Comment by oarbusi** [comment]
@EspenAlbert no haven't done it. To get that kind of info from the API, credentials need more permissions than we currently have, so I prefer to not increase permissions for this.
